src/main/scala/cromwell/engine/WorkflowManagerActor.scala

```diff
@@ -43,48 +45,28 @@ class WorkflowManagerActor extends Actor {
   private val workflowStore = new WriteOnceStore[WorkflowId, WorkflowActorRef]
   // This *should* be persisted
   private val workflowStates = TrieMap.empty[WorkflowId, WorkflowState]
-  private val listeners = MutableSet.empty[Promise[Unit]]
 
   def receive = LoggingReceive {
-    case SubmitWorkflow(wdl, inputs) =>
-      submitWorkflow(wdl, inputs) pipeTo sender
-
-    case WorkflowStatus(id) =>
-      sender ! workflowStates.get(id)
-
-    case NotifyCompletion(promise) =>
-      listeners += promise
-
-    case WorkflowOutputs(id) =>
-      // FIXME: What if the workflow isn't done? How best to handle?
-      workflowOutputs(id) pipeTo sender
-
-    case Shutdown =>
-      context.system.shutdown()
-
-    case WorkflowActor.Started =>
-      updateWorkflowState(sender(), WorkflowRunning)
-
-    case WorkflowActor.Done(outputs) =>
-      updateTerminalWorkflowState(WorkflowSucceeded)
-
-    case WorkflowActor.Failed(failures) =>
-      updateTerminalWorkflowState(WorkflowFailed)
-  }
-
-  private def updateTerminalWorkflowState(state: WorkflowState): Unit = {
-    listeners foreach { listener =>
-      if (!listener.isCompleted) listener.success(())
-    }
-    updateWorkflowState(sender(), state)
+    case SubmitWorkflow(wdl, inputs) => submitWorkflow(wdl, inputs) pipeTo sender
+    case WorkflowStatus(id) => sender ! workflowStates.get(id)
+    case Shutdown => context.system.shutdown()
+    // FIXME: What if the workflow isn't done? How best to handle?
+    case WorkflowOutputs(id) => workflowOutputs(id) pipeTo sender
+    case CurrentState(actor, state: WorkflowState) => updateWorkflowState(actor, state)
+    case Transition(actor, oldState, newState: WorkflowState) => updateWorkflowState(actor, newState)
+    case SubscribeToWorkflow(id) =>
+      val origSender = sender()
+      // FIXME: Silently fails if ID doesn't exist
+      workflowStore.toMap.get(id) foreach {_ ! SubscribeTransitionCallBack(origSender)}
   }
 
   private def workflowOutputs(id: WorkflowId): Future[Option[binding.WorkflowOutputs]] = {
     workflowStore.toMap.get(id) map workflowToOutputs getOrElse Future{None}
   }
 
-  private def workflowToOutputs(workflow: WorkflowActorRef): Future[Option[binding.WorkflowOutputs]] =
-    (workflow ? GetOutputs).mapTo[binding.WorkflowOutputs] map { Option(_) }
+  private def workflowToOutputs(workflow: WorkflowActorRef): Future[Option[binding.WorkflowOutputs]] = {
+    workflow.ask(GetOutputs).mapTo[binding.WorkflowOutputs] map {Option(_)}
```

Comment:
I see. This is an Option because `workflowOutputs` above wants to return an `Option` in the case where `id` doesn't exist. This doesn't seem like the right way to handle the whole situation (and I'm pretty sure I'm the original culprit!)